linkerd/core/src/main/scala/io/buoyant/linkerd/RequestEvaluator.scala

```diff
@@ -0,0 +1,165 @@
+  dtab: BaseDtab
+) extends SimpleFilter[Request, Response] {
+
+  private val EvaluatorHeaderName = "l5d-req-evaluate"
+  private val JsonCharSet = ";charset=UTF-8"
+
+  override def apply(
+    req: Request,
+    svc: Service[Request, Response]
+  ): Future[Response] = {
+
+    req.headerMap.get(EvaluatorHeaderName) match {
+      case None => svc(req)
+      case Some(_) =>
+
+        val identificationPath = DstPathCtx.current match {
+          case Some(dstPath) => dstPath.path
+          case None => Path.empty
+        }
+
+        val selectedEndpoint = endpoint.addr match {
+          case inetAddr: Address.Inet => inetAddr.addr.toString
+          case _ => ""
+        }
+
+        val lbSet = DstBoundCtx.current match {
+          case None => Addr.Neg
+          case Some(addrSet) => addrSet.name.addr.sample()
+        }
+
+        val addresses = lbSet match {
+          case Addr.Neg | Addr.Failed(_) | Addr.Pending => None
+          case Addr.Bound(a, _) => Some(
+            a.map { address: Address =>
+              address match {
+                case inetAddr: Address.Inet => inetAddr.addr.toString
+                case _ => ""
+              }
+            }
+          )
+        }
+
+        val dtreeF = namer.interpreter match {
+          case delegator: Delegator => delegator.delegate(dtab.dtab(), identificationPath)
+            .map(Some(_))
+          case _ => Future.None
+        }
+
+        dtreeF.map { dtree =>
+          val resp = Response()
+          val tree = RequestEvaluator.formatDTree(dtree, req.contentType, List.empty)
+          val evaluatedRequest = EvaluatedRequest(
+            Try(tree.last) match {
+              case Throw(_) => "Unknown identification"
+              case Return(id) => id
+            },
+            selectedEndpoint,
+            addresses,
+            tree
+          )
+
+          resp.contentType = req.contentType.getOrElse(MediaType.PlainText)
+          resp.contentString = RequestEvaluator
+            .writeContentString(resp.contentType, evaluatedRequest)
+          resp
+        }
+    }
+  }
+}
+
+object RequestEvaluator {
+
+  private val jsonMapper = Parser.jsonObjectMapper(Nil)
+    .registerModule(DelegationJsonCodec.mkModule())
+
+  val module: Stackable[ServiceFactory[Request, Response]] =
+    new Stack.Module3[EndpointAddr, RoutingFactory.BaseDtab, DstBindingFactory.Namer, ServiceFactory[Request, Response]] {
+
+      override def role: Stack.Role = Stack.Role("RequestEvaluator")
+
+      override def description: String = "Intercepts to respond with useful client destination info"
+
+      override def make(
+        endpoint: EndpointAddr,
+        dtab: BaseDtab,
+        interpreter: DstBindingFactory.Namer,
+        next: ServiceFactory[Request, Response]
+      ): ServiceFactory[Request, Response] = new RequestEvaluator(endpoint, interpreter, dtab)
+        .andThen(next)
+    }
+
+  def formatDTree(
+    dTree: Option[DelegateTree[Name.Bound]],
```

Comment:
all recursive calls to this method are called with a Some here.  Could this just be a `DelegateTree[Name.Bound]` and have the Option handling outside of the method?

Reply:
I changed the recursive code a little bit to be clearer. I also removed the use of `Option`.